AstuteSource/chasten
# Feature: ability to pull remote configuration files over http

closes https://github.com/AstuteSource/chasten/issues/17

This PR allows for users to specify a Url directive to a text file for the `--config` argument. Additionally, it tests all of the newly added features. It also moves around some functions to reduce ambiguity of the names of methods (e.g. moving things related to configuration to `configuration.py`)

Example usage:

```sh
chasten analyze test --config https://raw.githubusercontent.com/AstuteSource/chasten-configuration/master/config.yml --search-path .
```

Summary of all changes:

* `configuration.py`
  * `configuration.display_configuration_directory()` (moved)
    * Moved from `main.py` to `configuration.py`, because this is specific to configuration.
  * `configuration.validate_checks_file()` (new)
    * New method that extracts logic specific to checks file into a function
  * `configuration.validate_configuration_files()` (moved, modified)
    * Moved from `main.py` to `configuration.py`, because this is specific to configuration.
    * Modified to account for configuration files being `pathlib.Path` and `urllib3.Url`
  * `configuration.convert_configuration_text_to_yaml()` (new)
    * New method that breaks out functionality of larger method into a function.
  * `configuration.extract_configuration_details_from_config_url()` (new)
    * New method to distinguish extracting configuration details from URL versus Path
  * `configuration.extract_configuration_details_from_config_dir()` (new)
    * New method to distinguish extracting configuration details from URL versus Path
* `validate.py` (moved, modified)
  * `validate.validate_file()`
    * Moved from `main.py` to `validate.py`, because this is specific to validation
* `util.py`
  * `util.is_url()` (new)
    * New method to determine if string is valid URL
* `tests/`
  * `test_checks.py` (modified)
    * `JSON_SCHEMA_COUNT` modified to fill in rest of schema from canonical schema found in `validate.py`
  * `test_main.test_cli_analyze_url_config()` (new)
    * Makes sure CLI correctly handles simple URL config.
  * `test_main.test_cli_analyze_url_config_with_local_checks_file()` (new)
    * Confirm that using the command-line interface aborts execution when given a URL config that uses a local file path to specify checks files.
  * `test_main.test_cli_analyze_url_config_with_url_checks_file()` (new)
    * Confirm that using the command-line interface correctly handles a local config that references URL endpoints for each checks file.
  * `test_main.test_cli_analyze_local_config_with_url_and_local_checks_files()` (new)
    * Confirm that using the command-line interface correctly handles a local config that references a combination of URL endpoints and local files for each checks file.
  * `test_main.test_cli_analyze_url_config_with_url_and_local_checks_files()` (new)
    * Confirm that using the command-line interface aborts execution when given a URL config that references a combination of URL endpoints and local files for each checks file.
  * `test_util.test_is_url_correct()` (new)
    * Use Hypothesis to confirm that URLs are correctly recognized/unrecognized.
* `.chasten`
  * `.chasten/config_url_and_local_checks_files.yml` (new)
    * Local config used in testing.
  * `.chasten/config_url_and_checks_file.yml` (new)
    * Local config used in testing.
* `README.md` (edited)
  * Modified to indicate that URLs are allowed as valid configurations.
